src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/ConnectionImpl.java

```diff
@@ -264,24 +297,36 @@ private ClientSession connectAndAuthenticate() throws IOException {
         ConnectFuture connectionFuture = client.connect(this.credentials.getUsername(), this.host, this.port);
         connectionFuture.verify(this.timeoutMillis);
         session = connectionFuture.getSession();
-        addAuthentication();
-        AuthFuture auth = session.auth();
-        auth.verify(this.timeoutMillis);
+        try {
+            if (!SSHAuthenticator.newInstance(session, credentials).authenticate(TaskListener.NULL)) {
+                session.close(true);
+                throw new IOException("Authentication failed.");
+            }
+        } catch (InterruptedException e) {
+            throw new IOException(e);
+        }
+        //        addAuthentication();
+        //        AuthFuture auth = session.auth();
+        //        auth.verify(this.timeoutMillis);
         return session;
     }
 
     /** Initialize the SSH client. It reuses the client if it exists. */
     private void initClient() {
         if (client == null) {
-            client = SshClient.setUpDefaultClient();
+            ClientBuilder clientBuilder = ClientBuilder.builder();
+            if (getHostKeyVerifier() instanceof SSHApacheMinaLauncher.ServerHostKeyVerifierImpl hostKeyVerifier) {
+                clientBuilder.serverKeyVerifier(hostKeyVerifier.getServerKeyVerifier());
+            }
+            client = clientBuilder.build();
```

Comment:
I want to add the verifiers in a different PR, I have to though how to add them without making the code complex.

Reply:
well actually this make the code quite simple as just reusing current classes (just add a new method), not adding new classes and so will avoid complicated migration process for users (or even avoid us writing some migration tools)
once we remove trilead support we can remove the method `public abstract boolean verify(SlaveComputer computer, HostKey hostKey, TaskListener listener) throws Exception;` or deprecate it at least and make it doing nothing as not needed anymore.

Reply:
The issue is that we need to remove the old `slaves` package at some point, as it is a technical debt task. In any case, the migration should occur because the old launcher has to change to the new one, so there should be something to migrate.